codalab/lib/bundle_store.py

```diff
@@ -169,41 +259,413 @@ def upload(self, sources, follow_symlinks, exclude_patterns, git, unpack, remove
+
+    Q: What happens if a disk fails?
+    A: No plans for replication right now, but we can think about this later. In fact, replicating the bundle data is
+       probably something that makes sense long-term.
+
+    Q: What happens if we want to migrate away from one disk to another.
+    A: The hash ring takes care of this, and only the bundles on the disk we are migrating from will be moved.
+
+    Q: What happens when the disk a bundle is supposed to go to runs out of space?
+    A: Need to avoid this scenario by proactively adding new disk mounts when the time arises, and then
+       redistributing bundles when they need to be redistributed.
+    """
+
+    # Location where MultiDiskBundleStore keeps production and temp data relative to each disk
+    DATA_SUBDIRECTORY = 'mdata'
+    TEMP_SUBDIRECTORY = 'mtemp'
+
+    # Location for basic uploading
+    MISC_TEMP_SUBDIRECTORY = 'mtemp'
+
+    # Cleanup time before we are allowed to delete old bundles
+    DATA_CLEANUP_TIME = 60
+    TEMP_CLEANUP_TIME = 60 * 60
+
+    def __init__(self, codalab_home):
+        self.codalab_home = path_util.normalize(codalab_home)
+
+        self.mdisk = os.path.join(self.codalab_home, 'mdisk')
+        self.mtemp = os.path.join(self.codalab_home, MultiDiskBundleStore.MISC_TEMP_SUBDIRECTORY)
+
+        # Perform initialization first to ensure that directories will be populated
+        super(MultiDiskBundleStore, self).__init__()
+        nodes, _ = path_util.ls(self.mdisk)
+        self.ring = HashRing(nodes)
+
+    @require_disks
+    def upload(self, sources, follow_symlinks, exclude_patterns, git, unpack, remove_sources):
+        """
+        Performs a lookup in the hash ring before doing the upload
+        """
+        """
+        |sources|: specifies the locations of the contents to upload.  Each element is either a URL or a local path.
+        |follow_symlinks|: for local path(s), whether to follow (resolve) symlinks
+        |exclude_patterns|: for local path(s), don't upload these patterns (e.g., *.o)
+        |git|: for URL, whether |source| is a git repo to clone.
+        |unpack|: for each source in |sources|, whether to unpack it if it's an archive.
+        |remove_sources|: remove |sources|.
+
+        If |sources| contains one source, then the bundle contents will be that source.
+        Otherwise, the bundle contents will be a directory with each of the sources.
+        Exceptions:
+        - If |git|, then each source is replaced with the result of running 'git clone |source|'
+        - If |unpack| is True or a source is an archive (zip, tar.gz, etc.), then unpack the source.
+
+        Install the contents of the directory at |source| into
+        DATA_SUBDIRECTORY in a subdirectory named by a hash of the contents.
+
+        Return a (data_hash, metadata) pair, where the metadata is a dict mapping
+        keys to precomputed statistics about the new data directory.
+        """
+        to_delete = []
+
```

Comment:
Should figure out what disk to use first and then upload everything to the 'temp' directory of **that** disk.  That way, moving to the data directory is cheap and we can also fail early if that disk is broken or something.